addons/stock/report/report_replenishment.py

```diff
@@ -0,0 +1,423 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+from odoo.tools import format_datetime
+
+
+class ReplenishmentReport(models.AbstractModel):
+    _name = 'report.stock.report_product_product_replenishment'
+    _description = "Stock Replenishment Report"
+
+    @api.model
+    def _product_domain(self, product_template_ids, product_variant_ids):
+        domain = []
+        if product_template_ids:
+            domain += [('product_tmpl_id', 'in', product_template_ids)]
+        elif product_variant_ids:
+            domain += [('product_id', 'in', product_variant_ids)]
+        return domain
+
+    @api.model
+    def _get_report_values(self, docids, data=None):
+        docs = self._get_report_data(product_variant_ids=docids)
+        docargs = {
+            'data': data,
+            'doc_ids': docids,
+            'doc_model': 'product.product',
+            'docs': docs,
+        }
+        return docargs
+
+    @api.model
+    def _get_report_data(self, product_template_ids=False, product_variant_ids=False):
+        """ Return all the report data, which include report lines (documents,
+        quantity, expected date), quantity summary (On Hand Qty, Forecasted Qty
+        and Forcasted + Pending Qty) and the number of pending documents.
+
+        This method use the following context key:
+            - `warehouse`: will restrict the search move for this warehouse.
+              If missing, it will take the first company's warehouse.
+
+        :param product_template_ids: list of `product.product` ids.
+        :type product_template_ids: list
+        :param product_variant_ids: list of `product.template` ids.
+        :type product_variant_ids: list
+
+        :return: a dict with all the report data.
+        :rtype: dict
+        """
+        res = {
+            'product_templates': False,
+            'product_variants': False,
+        }
+        if self.user_has_groups('stock.group_stock_multi_warehouses') or len(self.env.companies) > 1:
+            # Add warehouse id in the context to reuse it to filter data to make
+            # the report warehouse specific.
+            warehouse_id = self.env.context.get('warehouse', False)
+            if warehouse_id:
+                warehouse = self.env['stock.warehouse'].browse(warehouse_id)
+            else:
+                warehouse = self.env['stock.warehouse'].search([
+                    ('company_id', '=', self.env.company.id)
+                ], limit=1)
```

Comment:
We never display the report for multi warehouse or multi companies?

Reply:
No, always only one warehouse/company at time.